src/kbucket.rs

```diff
@@ -115,34 +118,107 @@ impl KbucketTable {
                 }
             }
         }
-        SearchResult {
-            found: false,
-            bucket_index,
-            column_index: last_empty_index,
-        }
+        Search::Failure(bucket_index, last_empty_index)
     }
 
-    fn find_bucket_index(&self, identifier: Identifier) -> usize {
+    pub fn xor_bucket_index(&self, identifier: Identifier) -> usize {
         let x = U256::from(self.local_node_id);
         let y = U256::from(identifier);
         let xor_distance = x ^ y;
 
-        let bucket_index = MAX_BUCKETS - (xor_distance.leading_zeros() as usize);
-        println!(
-            "Xor distance leading zeros, {}",
-            xor_distance.leading_zeros()
-        );
-        println!("Bucket index for given key: {}", bucket_index);
-        bucket_index
+        MAX_BUCKETS - (xor_distance.leading_zeros() as usize)
     }
 }
 
-/// TODO:  Implement real deal tests!
-///
-/// Test find_node()      **Requires adding nodes to our table**
-/// Test search_table()
-/// Test add_node()
 #[cfg(test)]
 mod tests {
     use super::*;
+
+    fn mk_nodes(n: u8) -> (Node, Vec<Node>) {
+        let listen_addr = String::from("127.0.0.1").parse::<Ipv4Addr>().unwrap();
+        let port_start = 9000_u16;
+
+        let our_nodes: Vec<Node> = (0..n)
+            .map(|i| mk_node(&listen_addr, port_start, i))
+            .collect();
+
+        if let Some((local_node, remote_nodes)) = our_nodes.split_first() {
+            let remote_nodes = remote_nodes.to_vec();
+            (*local_node, remote_nodes)
+        } else {
+            panic!("Not enough nodes were created");
+        }
+    }
+
+    fn mk_node(listen_addr: &Ipv4Addr, port_start: u16, index: u8) -> Node {
+        let mut node_id = [0_u8; 32];
+        node_id[31] += index;
```

Comment:
you don't need to change this but I try to avoid `mut` where possible

instead you could do something like `let node_id = [index; 32];`

Reply:
Ahhhh...

I changed node_id from [index; 32] to this mut node_id so it's easier for me to reason about bucket distances.